aten/src/ATen/gen.py

```diff
@@ -368,10 +368,11 @@ def generate_storage_type_and_tensor(backend, density, declarations, per_op_regi
     if env['DeviceType'] == 'CPU':
         top_env['cpu_type_headers'].append(
             '#include <ATen/{}.h>'.format(env['Type']))
-    else:
-        assert env['DeviceType'] == 'CUDA'
+    elif env['DeviceType'] == 'CUDA':
         top_env['cuda_type_headers'].append(
             '#include <ATen/{}.h>'.format(env['Type']))
+    else:  # Allow gen.py to be called by out-of-tree devices
```

Comment:
why do you need this?

Reply:
1. I start with a dictionary of `kernel_name: fpga_dispatch_function`.
2. I copy the latest native_functions.yaml from PyTorch and add `FPGA: fpga_dispatch_funtion` under the ` dispatch entries`.  See mul example:

```
 func: mul.Tensor(Tensor self, Tensor other) -> Tensor
  use_c10_dispatcher: full
  variants: function, method
  dispatch:
    FPGA: mul  (I added this before calling your gen.py)
    CPU: mul
    CUDA: mul
    SparseCPU: mul_sparse
    SparseCUDA: mul_sparse
    MkldnnCPU: mkldnn_mul
```
3. I then call gen.py and it generates FPGAType.h and FPGAType.cpp and I copy these files to my extension repo.

I can then re-sync with pytorch every time native_functions.yaml changes.

Reply:
all you really need here is a way to call the CPU "mul.Tensor" implementation, right?  The rest is just native_function mechanics that don't really matter for your use case AFAICT.

idk if the dispatcher can do that, but I'm sure @smessmer or @ezyang knows.

Reply:
maybe some of the BackendSelect-style "redispatch with key" stuff could work.

Reply:
To clarify if you look at `torch/csrc/autograd/generated/RegistrationDeclarations.h`(after you build pytorch) that's all information XLA need to register ops in pytorch. Maybe we could reuse that for FPGA?

Reply:
> Can you paste an example of generated header/cpp files here?

Attaching files appears to be broken, so I'll link you to the files on gitlab.
[FPGAType.cpp](https://gitlab.com/pytorch-complex/vitis_kernels/-/blob/master/aten/src/ATen/FPGAType.cpp)
[FPGAType.h](https://gitlab.com/pytorch-complex/vitis_kernels/-/blob/master/aten/src/ATen/FPGAType.h)

I have a script called [gen.py](https://gitlab.com/pytorch-complex/vitis_kernels/-/blob/master/aten/src/ATen/gen.py) that appends `FPGA` to the list of backends in the Pytorch gen.py, which then generates `FPGAType.h` and `FPGAType.cpp`. 

> To clarify if you look at torch/csrc/autograd/generated/RegistrationDeclarations.h(after you build pytorch) that's all information XLA need to register ops in pytorch. Maybe we could reuse that for FPGA?

That file gives me enough information to dispatch at the device level, however the FPGAs that Xilinx markets towards Tensorflow and Caffe are always CPU + FPGA SoCs. The idea is the FPGA replaces the logic in the ATen/native/cpu folder, while the CPU is still used for the ATen/native folder. I've been operating under the impression that device-level dispatching was done this way.

Reply:
Hmmm I'm a bit confused. Do you mean FPGA plans to **only** support ops implemented through TensorIterator? 
What happens when we move new ops into ATen/native/cpu folder? Does that break your integration?

Reply:
@ailzhang

I'm confused here too. I don't understand what your criteria is for logic that goes into `ATen/native` vs. `ATen/native/cpu` and I'm worried about this.  I would use the following to separate these two:

### ATen/native/___.cpp
 - Any Assert statement
 - Anything statement that processes the entire tensor.
      - ie: like a call to another at:: function

### ATen/native/cpu/___.cpp (or ATen/native/fpga/___cpp
- Only processes tensor data
- Simple "Bytes-In, Bytes-Out" interface with few optional or keyword arguments.
- Tensor data is consumed:
    - All at once in eager mode.
    - Using a streaming interface (Tensor is process one-byte at a time).